pep-perennial.rst

```diff
@@ -97,6 +97,17 @@ profiles in auditwheel. At least initially, they will likely be similar to
 the list for manylinux2010 (:pep:`571`), and based on library versions in
 newer versions of CentOS.
 
+The overall goal is to ensure that if a wheel is tagged as
+``manylinux_glibc_2_Y``, then users can be reasonably confident that this wheel
+will work in any real-world linux-based python environment that uses
+``glibc 2.Y`` or later and matches the other wheel compatibility tags.
+For example, this includes making sure that the wheel only uses symbols that
+are available in the oldest supported glibc, and doesn't rely on the system to
+provide any libraries that aren't universally available.
+But the exact details of auditwheel's checks will evolve over time as the Linux
+distribution landscape changes and as we learn more about real-world
+compatibility pitfalls.
+
```

Comment:
I haven't done an exhaustive investigation here, but I want to raise the issue of source and binary compatibility for C++ code right now: this should be explicitly addressed in the PEP.  There is a catch-22 here:

* New features are constantly being added to C++ the language; new code has a good chance of failing to compile with an old compiler, and contrariwise old code has a good chance of failing to compile with a new compiler.  Each new version of GCC adds many new symbols to `libstdc++.so.6`.
* However, _bundling libstdc++ in wheels is not safe_, because it has internal global state.  If there is more than one copy of libstdc++ loaded into a process, the behavior of the entire program becomes undefined, in the sense in which the C and C++ standards use that term.

This _may_ mean that it is necessary to write the version of the C++ compiler into the name of the wheel, _and_ to say that `manylinux_glibc_X_g++_Y` is not coinstallable with `manylinux_glibc_X_g++_Z` where Y ≠ Z.  It's possible that there's a cleverer option, but I am not sure what it would be.

Reply:
Huh, is that true even if I have two libraries that statically link their own copies of libstdc++ (or that dynamically load libstdc++ into disjoint ELF namespaces), and whose public APIs are entirely C based?

Anyway, the way we handle this right now is basically that if you want to make a wheel that targets, say, 2010-era Linux, then you dynamically link to the system `libstdc++` and you assume it only contains 2010-era C++ symbols.

This is somewhat restrictive, but less restrictive than you would think. Redhat distributes newer compilers that contain clever linker scripts, that use the system symbols whenever they're available, and for any unavailable symbols it statically links them into the binary. These are the compilers we provide in the manylinux images.

And anyway, I don't think we can hope to do any better. For manylinux wheels, we generally assume that they might have to coexist with arbitrary other extensions, built in arbitrary ways. For example, there might be packages that the user compiled themselves using their regular distro compiler, and that link to the system libstdc++. According to your rule, this means that we can never vendor libstdc++ at all; we have to use the system libstdc++. And if we're stuck using the system libstdc++, then the glibc version pretty much tells us what that is, so there's not much point in putting it in the wheel filename.

Reply:
>> If there is more than one copy of libstdc++ loaded into a process, the behavior of the entire program becomes undefined
>
> is that true even if I have two libraries that statically link their own copies of libstdc++ (or that dynamically load libstdc++ into disjoint ELF namespaces), and whose public APIs are entirely C based?

Yes.  The root issue is that each copy of the library assumes it has the _only_ copy of a global data structure (such as: tables of exception handling information, tables of RTTI information, the `cin` object, the canonical empty `std::string`, etc).  Malfunctions I've personally witnessed include `std::terminate` and/or `std::unexpected` getting called in a program that should never have called them, `typeid` invariants being violated, and segmentation faults upon attempting to write to a `std::string` object that should have been mutable.  I wouldn't be surprised if the problem with `call_once` that came up in one of the precursor discussions to this PEP could be traced to the same issue.

Some of the problems could be mitigated by ensuring that the Python core interpreter executable and/or `libpython.so` were linked against `libgcc_s.so.1`; to make this fully reliable, interpreter initialization would need to call a function that is only defined in that library, ideally before any threads have been created.  (This would also fix some of the problems currently being muddled together in https://bugs.python.org/issue18748 .)  But it wouldn't be a complete solution and I'm not sure what a good function to call would be.

> if we're stuck using the system libstdc++, then the glibc version pretty much tells us what that is

This may be true for CentOS but I know some other distributions (e.g. Debian, Arch) do not update GCC on the same schedule as glibc, so it's not perfect.

Documenting a specific gcc and g++ version to be used to build wheels conforming to each rev of manylinux would probably be enough to address the problem, as long as CPython itself continues not to contain any C++ code.  However, it was my impression that the idea of perennial manylinux was to avoid having to research and document such things?

Reply:
> Yes. The root issue is that each copy of the library assumes it has the only copy of a global data structure (such as: tables of exception handling information, tables of RTTI information, the cin object, the canonical empty std::string, etc).

Sure, I get that part. What I'm missing is what libstdc++ does to thwart all the stuff that linkers do to try to make this situation work anyway. For example if I make my own `nonstd::EMPTY_STRING` and it gets loaded twice into different ELF namespaces that only communicate via C, it's generally fine – each namespace consistently uses its own copy of the singleton.

Anyway, we don't recommend vendoring libstdc++ so this is mostly academic curiosity, and also in case it gives some hint to the mysterious `call_once` issue.

> This may be true for CentOS but I know some other distributions (e.g. Debian, Arch) do not update GCC on the same schedule as glibc, so it's not perfect.

That's fine – we don't care about the leading edge, only the trailing edge, and the trailing edge is much more stable and boring. So e.g. consider a `manylinux_23`, targeting systems with glibc 2.23 and newer. glibc 2.23 was released in 2016. If we look around at all the distros currently shipping with glibc 2.23 or later, and compute the minimum libstdc++ version, it's probably going to be something from 2016-ish. The situation where we'd get in trouble is if suddenly a *new* distro appears that's shipping an *older* version of libstdc++ than any of the distros we already looked at – so you'd need a new 2019 distro that just started shipping a libstdc++ from 2015-ish. That never happens in practice.

Sure, this means that it's actually impossible to know what the definition of manylinux_X is until some time after glibc 2.X has been out in the wild – the standard says that manylinux_X wheels have to be compatible with all distros shipping glibc 2.X, and that's not determined until all those distros have shipped. But that's unavoidable, and not really a problem since our goal is to achieve broad real-world compatibility, not generate wheels that only work on Arch.

>  However, it was my impression that the idea of perennial manylinux was to avoid having to research and document such things?

No, the goal is to do the same research and documentation, but decouple it from the PEP review cycle and pip release cycle.

Reply:
> > each copy of the library assumes it has the only copy of a global data structure
> What I'm missing is what libstdc++ does to thwart all the stuff that linkers do to try to make this situation work anyway

The uniqueness requirement applies across the entire _address space_.  ELF namespaces don't do a thing to help; in fact they may make the situation worse, by preventing symbol deduplication from happening.

With `typeid` and exceptions (which rely on typeid comparison) the problem is that typeid objects are compared by address.  This means that if you have a `std::runtime_error` thrown out of code using libstdc++ copy 1, code using libstdc++ copy 2 will be unable to catch it as a `std::runtime_error`. In fact, I think it can only catch it using `catch (...)`.

I don't know specifically what the problem is with `call_once` or the empty `std::string` but I believe it is something along the same lines.

Reply:
Maybe unrelated, but what do the current PEP's say about manylinux1 and manylinux2010 libraries being loaded at the same time? I am guessing if they both link libstdc++, they may present a similar issue.
Was this issue not considered before?

Reply:
@gunan I'm not sure how I managed to miss your comment until now.  You're right, this is a problem right now because, according to current policy, `libstdc++` is to be bundled in wheels, so a manylinux1 and a manylinux2010 wheel that both use C++ would each include a different version of the library.  But if `libstdc++` is added to the don't-bundle list, which I think is a necessary part of the fix anyway, then both wheels will pick it up from the system, and it should be OK as long as it's new enough.  (Requiring a specific version of G++ to be used is just to control what "new enough" means.  I think.)

Reply:
`libstdc++` has always been on the "don't bundle" list. However, we have recommended the use of the RH-maintained "devtoolset" compilers, which allow the use of new C++ features on old systems by selectively statically linking *pieces* of newer libstdc++ into the generated binaries, via some linker script magic.